File: js/base/SFEventHandler.js
;
/*!
 *
 *  SpriteForest JavaScript Library v1.0.0
 *  http://github.com/moky/SpriteForest/js/
 *
 *  Copyright (c) 2014 moKy at slanissue.com
 *  Released under the MIT license
 *  http://en.wikipedia.org/wiki/MIT_License
 *
 */
//
//  SFEventHandler.js
//  SpriteForest-HTML5
//
//  Created by moKy on Feb. 28, 2014.
//

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	//
	//  constructor
	//
	var EventHandler = function() {
		this._eventTree = null;
		this._actionTree = null;
	};
	
	//
	//  static functions
	//
	var create = function(eventTree, actionTree) {
		var handler = new sf.EventHandler();
		handler.init(eventTree, actionTree);
		return handler;
	};
	
	//
	//  instance functions
	//
	var init = function(eventTree, actionTree) {
		this._eventTree = eventTree;
		this._actionTree = actionTree;
		
		this.snapshot(); // for restore 'Target'
		
		return true;
	};
	
	var getTarget = function(dict, node) {
		var target = dict['Target'];
		if (!target) {
			target = dict['ss_Target'];
		}
		if (!target) {
			return node;
		}
		
		if (typeof target === 'string') {
			return sf.Node.getTarget(target, node);
		}
		return target;
	};
	
	var runAction = function(dict, node) {
		dict['Target'] = node;
		// create action
		var action = sf.Action.create(dict);
		if (!action) {
			//cn.error(dict);
			return false;
		}
		var ccAction = action.ccAction();
		if (ccAction instanceof cc.Action) {
			// run action
			node.runAction(ccAction);
			return true;
		}
		cn.error(action);
		return false;
	};
	
	var doEvent = function(eventName, node) {
		if (!this._eventTree || !node) {
			//cn.log(eventName);
			return false;
		}
		if (!this._eventTree) {
			return false;
		}
		
		var tasks = this._eventTree[eventName];
		if (!tasks) {
			return false;
		}
		
		var flag = false;
		
		for (var i = 0; i < tasks.length; ++i) {
			var task = tasks[i];
			var target = getTarget(task, node);
			var worker = this;
			
			if (target instanceof sf.Sprite) {
				worker = target.getEventDelegate();
			}
			
			var name = task['Name'];
			if (!worker) {
				cn.error('could not happen!');
			} else if (worker.doEvent(name, target)) {
				flag = true;
			} else if (worker.doAction(name, target)) {
				// combined action
				flag = true;
			} else {
				// cc action
				flag = runAction(task, target);
			}
		}
		
		return flag;
	};
	
	var doAction = function(actionName, node) {
		if (!this._eventTree || !node) {
			//cn.log(eventName);
			return false;
		}
		if (!this._actionTree) {
			return false;
		}
		
		var dict = this._actionTree[actionName];
		if (!dict) {
			return false;
		}
		
		var target = getTarget(dict, node);
		return runAction(dict, target);
	};
	
	var snapshotArray = function(array) {
		if (array)
		for (var i = array.length - 1; i >= 0; --i) {
			snapshotDictionary(array[i]);
		}
	};
	var restoreArray = function(array) {
		if (array)
		for (var i = array.length - 1; i >= 0; --i) {
			restoreDictionary(array[i]);
		}
	};
	var snapshotDictionary = function(dict) {
		if (dict)
		for (var key in dict) {
			var obj = dict[key];
			if (key === 'Target' && typeof obj === 'string') {
				dict['ss_Target'] = obj; // snapshot
			} else if (obj instanceof Array) {
				snapshotArray(obj);
			} else if (typeof obj === 'object') {
				snapshotDictionary(obj);
			}
		}
	};
	var restoreDictionary = function(dict) {
		if (dict)
		for (var key in dict) {
			var obj = dict[key];
			if (key === 'Target' && typeof obj !== 'string') {
				obj = dict['ss_Target'];
				if (obj) {
					dict['Target']; // restore
					dict['ss_Target'] = null;
				} else {
					dict['Target'] = null;
				}
			} else if (obj instanceof cc.Class) {
				// do nothing
			} else if (obj instanceof Array) {
				restoreArray(obj);
			} else if (typeof obj === 'object') {
				restoreDictionary(obj);
			}
		}
	};
	
	var restore = function() {
		restoreDictionary(this._eventTree);
		restoreDictionary(this._actionTree);
	};
	
	var snapshot = function() {
		snapshotDictionary(this._eventTree);
		snapshotDictionary(this._actionTree);
	};
	
	//--------------------------------------------------------------------------
	
	var prop = {
		ctor: EventHandler,
		
		init: init,
		
		doEvent: doEvent,
		doAction: doAction,
		
		restore: restore,
		snapshot: snapshot,
		
		// properties
		_eventTree: null,
		_actionTree: null,
	};
	
	// extends
	sf.EventHandler = cc.Class.extend(prop);
	
	// static
	sf.EventHandler.create = create;
	
}(SpriteForest);

File: objc/base/SFEventHandler.m
//
//  SFEventHandler.m
//  SpriteForest2
//
//  Created by Moky on 12-9-30.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#import "SFEventHandler.h"

#import "SFSprite.h"
#import "SFAction.h"

@implementation SFEventHandler

@synthesize eventTree = eventTree_;
@synthesize actionTree = actionTree_;

+ (id) handlerWithEventTree:(NSDictionary *)eTree actionTree:(NSDictionary *)aTree
{
	return [[[self alloc] initWithEventTree:eTree actionTree:aTree] autorelease];
}

- (id) initWithEventTree:(NSDictionary *)eTree actionTree:(NSDictionary *)aTree
{
	if ((self = [self init]))
	{
		if ([eTree isKindOfClass:[NSDictionary class]])
		{
			self.eventTree = eTree;
		}
		if ([aTree isKindOfClass:[NSDictionary class]])
		{
			self.actionTree = aTree;
		}
	}
	return self;
}

- (id) init
{
	if ((self = [super init]))
	{
		self.eventTree = nil;
		self.actionTree = nil;
	}
	return self;
}

- (void) dealloc
{
	[eventTree_ release];
	[actionTree_ release];
	
	[super dealloc];
}

#pragma mark -

- (BOOL) doAction:(NSString *)action node:(CCNode *)node
{
	if (![action isKindOfClass:[NSString class]])
	{
		CNLog(@"invalid action name: %@", action);
		return NO;
	}
	if (![node isKindOfClass:[CCNode class]])
	{
		CNLog(@"invalid node: %@", node);
		return NO;
	}
	
	NSDictionary * dict = [actionTree_ objectForKey:action];
	if (![dict isKindOfClass:[NSDictionary class]])
	{
		CNLog(@"no such action: %@", action);
		return NO;
	}
	
	// replace 'Target'
	CCNode * target = [node getTarget:[dict objectForKey:@"Target"]];
	if (![target isKindOfClass:[CCNode class]])
	{
		CNLog(@"cannot find target, dict = %@", dict);
		return NO;
	}
	
	BOOL flag = NO;
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	
	dict = [NSMutableDictionary dictionaryWithDictionary:dict];
	[(NSMutableDictionary *)dict setObject:target forKey:@"Target"];
	CCAction * act = [[SFAction getInstance:dict] ccAction];
	if ([act isKindOfClass:[CCAction class]])
	{
		// run it
		[target runAction:act];
		flag = YES;
	}
	else
	{
		CNLog(@"failed to create action, dict = %@, act = %@", dict, act);
	}
	
	[pool release];
	return flag;
}

- (BOOL) doEvent:(NSString *)event node:(CCNode *)node
{
	NSArray * tasks = [eventTree_ objectForKey:event];
	if (![tasks isKindOfClass:[NSArray class]])
	{
		//CNLog(@"no such event: %@", event);
		return NO;
	}
	
	BOOL flag = NO;
	
	for (NSDictionary * task in tasks)
	{
		if (![task isKindOfClass:[NSDictionary class]])
		{
			CNLog(@"invalid task (%@) in this event (%@)", task, event);
			continue;
		}
		CCNode * target = [node getTarget:[task objectForKey:@"Target"]];
		if (![target isKindOfClass:[CCNode class]])
		{
			CNLog(@"cannot find target, task = %@", task);
			continue;
		}
		
		id<SFEventDelegate> worker = self;
		if ([target isKindOfClass:[SFSprite class]])
		{
			worker = [(SFSprite *)target eventHandler];
		}
		
		NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
		
		NSString * name = [task objectForKey:@"Name"];
		if (![name isKindOfClass:[NSString class]])
		{
			CNLog(@"ERROR: invalid name = %@", name);
		}
		else if ([worker doEvent:name node:target])
		{
			flag = YES;
		}
		else if ([worker doAction:name node:target])
		{
			// combined action
			flag = YES;
		}
		else
		{
			// cc action
			task = [NSMutableDictionary dictionaryWithDictionary:task];
			[(NSMutableDictionary *)task setObject:target forKey:@"Target"];
			CCAction * act = [[SFAction getInstance:task] ccAction];
			if ([act isKindOfClass:[CCAction class]])
			{
				[target runAction:act];
				flag = YES;
			}
			else
			{
				CNLog(@"failed to create action, task = %@, act = %@", task, act);
			}
		}
		
		[pool release];
	} // EOF 'for (NSDictionary * task in tasks)'
	
	return flag;
}

@end
